pkg/ccpatch/patch_userdata.go

```diff
@@ -0,0 +1,89 @@
+package ccpatch
+
+import (
+	"encoding/base64"
+	"errors"
+	"strings"
+	"unicode"
+
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	// ErrNoCloudConfig is returned when no cloud-config header is found
+	ErrNoCloudConfig = errors.New("no cloud-config header found")
+	// ErrMalformedData is returned when the cloud-init data is malformed
+	ErrMalformedData = errors.New("malformed cloud-init data")
+)
+
+// CloudInitPatcher is a helper to patch cloud-init userdata
+type CloudInitPatcher struct {
+	raw     string
+	decoded string
+	data    map[string]interface{}
+}
+
+// NewCloudInitPatcher returns a new CloudInitPatcher instance
+// from a base64 encoded string
+func NewCloudInitPatcher(raw string) (*CloudInitPatcher, error) {
+	byt, err := base64.StdEncoding.DecodeString(raw)
+	if err != nil {
+		return nil, ErrNoCloudConfig
```

Comment:
do we want to wrap and forward the errors so the user knows that decoding failed?